File: mRemoteV1/UI/UI.Window.Update.vb
Imports System.ComponentModel
Imports System.IO
Imports System.Net
Imports mRemote3G.App
Imports WeifenLuo.WinFormsUI.Docking

Namespace UI

    Namespace Window
        Public Class Update
            Inherits Base

#Region "Public Methods"

            Public Sub New(panel As DockContent)
                WindowType = Type.Update
                DockPnl = panel
                InitializeComponent()
                Runtime.FontOverride(Me)
            End Sub

#End Region

#Region "Form Stuff"

            Private Sub Update_Load(sender As Object, e As EventArgs) Handles Me.Load
                ApplyLanguage()
                CheckForUpdate()
            End Sub

            Private Sub ApplyLanguage()
                Text = Language.Language.strMenuCheckForUpdates
                TabText = Language.Language.strMenuCheckForUpdates
                btnCheckForUpdate.Text = Language.Language.strCheckForUpdate
                btnDownload.Text = Language.Language.strDownloadAndInstall
                lblChangeLogLabel.Text = Language.Language.strLabelChangeLog
                lblInstalledVersion.Text = Language.Language.strVersion
                lblInstalledVersionLabel.Text = String.Format("{0}:", Language.Language.strCurrentVersion)
                lblLatestVersion.Text = Language.Language.strVersion
                lblLatestVersionLabel.Text = String.Format("{0}:", Language.Language.strAvailableVersion)
            End Sub

            Private Sub btnCheckForUpdate_Click(sender As Object, e As EventArgs) Handles btnCheckForUpdate.Click
                CheckForUpdate()
            End Sub

            Private Sub btnDownload_Click(sender As Object, e As EventArgs) Handles btnDownload.Click
                DownloadUpdate()
            End Sub

            Private Sub pbUpdateImage_Click(sender As Object, e As EventArgs) Handles pbUpdateImage.Click
                Dim linkUri = TryCast(pbUpdateImage.Tag, Uri)
                If linkUri Is Nothing OrElse linkUri.IsFile Or linkUri.IsUnc Or linkUri.IsLoopback Then Return
                Process.Start(linkUri.ToString())
            End Sub

#End Region

#Region "Private Fields"

            Private _appUpdate As App.Update
            Private _isUpdateDownloadHandlerDeclared As Boolean = False

#End Region

#Region "Private Methods"

            Private Sub CheckForUpdate()
                If _appUpdate Is Nothing Then
                    _appUpdate = New App.Update
                ElseIf _appUpdate.IsGetUpdateInfoRunning Then
                    Return
                End If

                'lblStatus.Text = Language.strUpdateCheckingLabel
                lblStatus.Text = "Update Check is currently disabled."
                lblStatus.ForeColor = SystemColors.WindowText
                lblLatestVersionLabel.Visible = False
                lblInstalledVersion.Visible = False
                lblInstalledVersionLabel.Visible = False
                lblLatestVersion.Visible = False
                btnCheckForUpdate.Visible = False
                pnlUpdate.Visible = False

                Return

                AddHandler _appUpdate.GetUpdateInfoCompletedEvent, AddressOf GetUpdateInfoCompleted

                _appUpdate.GetUpdateInfoAsync()
            End Sub

            Private Sub GetUpdateInfoCompleted(sender As Object, e As AsyncCompletedEventArgs)
                If InvokeRequired Then
                    Dim myDelegate As New AsyncCompletedEventHandler(AddressOf GetUpdateInfoCompleted)
                    Invoke(myDelegate, New Object() {sender, e})
                    Return
                End If

                Try
                    RemoveHandler _appUpdate.GetUpdateInfoCompletedEvent, AddressOf GetUpdateInfoCompleted

                    lblInstalledVersion.Text = My.Application.Info.Version.ToString
                    lblInstalledVersion.Visible = True
                    lblInstalledVersionLabel.Visible = True
                    btnCheckForUpdate.Visible = True

                    If e.Cancelled Then Return
                    If e.Error IsNot Nothing Then Throw e.Error

                    If _appUpdate.IsUpdateAvailable() Then
                        lblStatus.Text = Language.Language.strUpdateAvailable
                        lblStatus.ForeColor = Color.OrangeRed
                        pnlUpdate.Visible = True

                        Dim updateInfo As App.Update.UpdateInfo = _appUpdate.CurrentUpdateInfo
                        lblLatestVersion.Text = updateInfo.Version.ToString
                        lblLatestVersionLabel.Visible = True
                        lblLatestVersion.Visible = True

                        If _
                            updateInfo.ImageAddress Is Nothing OrElse
                            String.IsNullOrEmpty(updateInfo.ImageAddress.ToString()) Then
                            pbUpdateImage.Visible = False
                        Else
                            pbUpdateImage.ImageLocation = updateInfo.ImageAddress.ToString()
                            pbUpdateImage.Tag = updateInfo.ImageLinkAddress
                            pbUpdateImage.Visible = True
                        End If

                        AddHandler _appUpdate.GetChangeLogCompletedEvent, AddressOf GetChangeLogCompleted
                        _appUpdate.GetChangeLogAsync()

                        btnDownload.Focus()
                    Else
                        lblStatus.Text = Language.Language.strNoUpdateAvailable
                        lblStatus.ForeColor = Color.ForestGreen

                        If _appUpdate.CurrentUpdateInfo IsNot Nothing Then
                            Dim updateInfo As App.Update.UpdateInfo = _appUpdate.CurrentUpdateInfo
                            If updateInfo.IsValid And updateInfo.Version IsNot Nothing Then
                                lblLatestVersion.Text = updateInfo.Version.ToString
                                lblLatestVersionLabel.Visible = True
                                lblLatestVersion.Visible = True
                            End If
                        End If
                    End If
                Catch ex As Exception
                    lblStatus.Text = Language.Language.strUpdateCheckFailedLabel
                    lblStatus.ForeColor = Color.OrangeRed

                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strUpdateCheckCompleteFailed, ex)
                End Try
            End Sub

            Private Sub GetChangeLogCompleted(sender As Object, e As AsyncCompletedEventArgs)
                If InvokeRequired Then
                    Dim myDelegate As New AsyncCompletedEventHandler(AddressOf GetChangeLogCompleted)
                    Invoke(myDelegate, New Object() {sender, e})
                    Return
                End If

                Try
                    RemoveHandler _appUpdate.GetChangeLogCompletedEvent, AddressOf GetChangeLogCompleted

                    If e.Cancelled Then Return
                    If e.Error IsNot Nothing Then Throw e.Error

                    txtChangeLog.Text = _appUpdate.ChangeLog
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strUpdateGetChangeLogFailed, ex)
                End Try
            End Sub

            Private Sub DownloadUpdate()
                Try
                    btnDownload.Enabled = False
                    prgbDownload.Visible = True
                    prgbDownload.Value = 0

                    If _isUpdateDownloadHandlerDeclared = False Then
                        AddHandler _appUpdate.DownloadUpdateProgressChangedEvent,
                            AddressOf DownloadUpdateProgressChanged
                        AddHandler _appUpdate.DownloadUpdateCompletedEvent, AddressOf DownloadUpdateCompleted
                        _isUpdateDownloadHandlerDeclared = True
                    End If

                    _appUpdate.DownloadUpdateAsync()
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strUpdateDownloadFailed, ex)
                End Try
            End Sub

#End Region

#Region "Events"

            Private Sub DownloadUpdateProgressChanged(sender As Object, e As DownloadProgressChangedEventArgs)
                prgbDownload.Value = e.ProgressPercentage
            End Sub

            Private Sub DownloadUpdateCompleted(sender As Object, e As AsyncCompletedEventArgs)
                Try
                    btnDownload.Enabled = True
                    prgbDownload.Visible = False

                    If e.Cancelled Then Return
                    If e.Error IsNot Nothing Then Throw e.Error

                    If _
                        MessageBox.Show(Language.Language.strUpdateDownloadComplete,
                                        Language.Language.strMenuCheckForUpdates, MessageBoxButtons.OKCancel,
                                        MessageBoxIcon.Warning) = DialogResult.OK Then
                        Runtime.Shutdown.Quit(_appUpdate.CurrentUpdateInfo.UpdateFilePath)
                        Return
                    Else
                        File.Delete(_appUpdate.CurrentUpdateInfo.UpdateFilePath)
                    End If
                Catch ex As Exception
                    Runtime.MessageCollector.AddExceptionMessage(Language.Language.strUpdateDownloadCompleteFailed, ex)
                End Try
            End Sub

#End Region
        End Class
    End Namespace

End Namespace

File: mRemoteV1/CS/UI/UI.Window.Update.cs
using System.Collections.Generic;
using System;
using AxWFICALib;
using System.Drawing;
using System.Diagnostics;
using System.Data;
using AxMSTSCLib;
using Microsoft.VisualBasic;
using System.Collections;
using System.Windows.Forms;
using System.ComponentModel;
using mRemoteNG.My;
using WeifenLuo.WinFormsUI.Docking;
using System.IO;
using mRemoteNG.App;


namespace mRemoteNG.UI.Window
{
	public partial class Update : Base
	{
        #region Public Methods
		public Update(DockContent panel)
		{
			WindowType = Type.Update;
			DockPnl = panel;
			InitializeComponent();
			FontOverride(this);
		}
        #endregion
				
        #region Form Stuff
		public void Update_Load(object sender, EventArgs e)
		{
			ApplyLanguage();
			CheckForUpdate();
		}
				
		private void ApplyLanguage()
		{
			Text = Language.strMenuCheckForUpdates;
			TabText = Language.strMenuCheckForUpdates;
			btnCheckForUpdate.Text = Language.strCheckForUpdate;
			btnDownload.Text = Language.strDownloadAndInstall;
			lblChangeLogLabel.Text = Language.strLabelChangeLog;
			lblInstalledVersion.Text = Language.strVersion;
			lblInstalledVersionLabel.Text = string.Format("{0}:", Language.strCurrentVersion);
			lblLatestVersion.Text = Language.strVersion;
			lblLatestVersionLabel.Text = string.Format("{0}:", Language.strAvailableVersion);
		}
				
		public void btnCheckForUpdate_Click(System.Object sender, EventArgs e)
		{
			CheckForUpdate();
		}
				
		public void btnDownload_Click(System.Object sender, EventArgs e)
		{
			DownloadUpdate();
		}
				
		public void pbUpdateImage_Click(System.Object sender, EventArgs e)
		{
			Uri linkUri = pbUpdateImage.Tag as Uri;
			if (linkUri == null || linkUri.IsFile || linkUri.IsUnc || linkUri.IsLoopback)
			{
				return ;
			}
			Process.Start(linkUri.ToString());
		}
        #endregion
				
        #region Private Fields
		private App.Update _appUpdate;
		private bool _isUpdateDownloadHandlerDeclared = false;
        #endregion
				
        #region Private Methods
		private void CheckForUpdate()
		{
			if (_appUpdate == null)
			{
				_appUpdate = new App.Update();
				_appUpdate.Load += _appUpdate.Update_Load;
			}
			else if (_appUpdate.IsGetUpdateInfoRunning)
			{
				return ;
			}
					
			lblStatus.Text = Language.strUpdateCheckingLabel;
			lblStatus.ForeColor = SystemColors.WindowText;
			lblLatestVersionLabel.Visible = false;
			lblInstalledVersion.Visible = false;
			lblInstalledVersionLabel.Visible = false;
			lblLatestVersion.Visible = false;
			btnCheckForUpdate.Visible = false;
			pnlUpdate.Visible = false;
					
			_appUpdate.GetUpdateInfoCompletedEvent += GetUpdateInfoCompleted;
					
			_appUpdate.GetUpdateInfoAsync();
		}
				
		private void GetUpdateInfoCompleted(object sender, AsyncCompletedEventArgs e)
		{
			if (InvokeRequired)
			{
				AsyncCompletedEventHandler myDelegate = new AsyncCompletedEventHandler(GetUpdateInfoCompleted);
				Invoke(myDelegate, new object[] {sender, e});
				return ;
			}
					
			try
			{
				_appUpdate.GetUpdateInfoCompletedEvent -= GetUpdateInfoCompleted;
						
				lblInstalledVersion.Text = System.Convert.ToString(Application.Info.Version.ToString());
				lblInstalledVersion.Visible = true;
				lblInstalledVersionLabel.Visible = true;
				btnCheckForUpdate.Visible = true;
						
				if (e.Cancelled)
				{
					return ;
				}
				if (e.Error != null)
				{
					throw (e.Error);
				}
						
				if (_appUpdate.IsUpdateAvailable())
				{
					lblStatus.Text = Language.strUpdateAvailable;
					lblStatus.ForeColor = Color.OrangeRed;
					pnlUpdate.Visible = true;
							
					App.Update.UpdateInfo updateInfo = _appUpdate.CurrentUpdateInfo;
					lblLatestVersion.Text = updateInfo.Version.ToString();
					lblLatestVersionLabel.Visible = true;
					lblLatestVersion.Visible = true;
							
					if (updateInfo.ImageAddress == null || string.IsNullOrEmpty(updateInfo.ImageAddress.ToString()))
					{
						pbUpdateImage.Visible = false;
					}
					else
					{
						pbUpdateImage.ImageLocation = updateInfo.ImageAddress.ToString();
						pbUpdateImage.Tag = updateInfo.ImageLinkAddress;
						pbUpdateImage.Visible = true;
					}
							
					_appUpdate.GetChangeLogCompletedEvent += GetChangeLogCompleted;
					_appUpdate.GetChangeLogAsync();
							
					btnDownload.Focus();
				}
				else
				{
					lblStatus.Text = Language.strNoUpdateAvailable;
					lblStatus.ForeColor = Color.ForestGreen;
							
					if (_appUpdate.CurrentUpdateInfo != null)
					{
						App.Update.UpdateInfo updateInfo = _appUpdate.CurrentUpdateInfo;
						if (updateInfo.IsValid && updateInfo.Version != null)
						{
							lblLatestVersion.Text = updateInfo.Version.ToString();
							lblLatestVersionLabel.Visible = true;
							lblLatestVersion.Visible = true;
						}
					}
				}
			}
			catch (Exception ex)
			{
				lblStatus.Text = Language.strUpdateCheckFailedLabel;
				lblStatus.ForeColor = Color.OrangeRed;
						
				Runtime.MessageCollector.AddExceptionMessage(Language.strUpdateCheckCompleteFailed, ex);
			}
		}
				
		private void GetChangeLogCompleted(object sender, System.ComponentModel.AsyncCompletedEventArgs e)
		{
			if (InvokeRequired)
			{
				AsyncCompletedEventHandler myDelegate = new AsyncCompletedEventHandler(GetChangeLogCompleted);
				Invoke(myDelegate, new object[] {sender, e});
				return ;
			}
					
			try
			{
				_appUpdate.GetChangeLogCompletedEvent -= GetChangeLogCompleted;
						
				if (e.Cancelled)
				{
					return ;
				}
				if (e.Error != null)
				{
					throw (e.Error);
				}
						
				txtChangeLog.Text = _appUpdate.ChangeLog;
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(Language.strUpdateGetChangeLogFailed, ex);
			}
		}
				
		private void DownloadUpdate()
		{
			try
			{
				btnDownload.Enabled = false;
				prgbDownload.Visible = true;
				prgbDownload.Value = 0;
						
				if (_isUpdateDownloadHandlerDeclared == false)
				{
					_appUpdate.DownloadUpdateProgressChangedEvent += DownloadUpdateProgressChanged;
					_appUpdate.DownloadUpdateCompletedEvent += DownloadUpdateCompleted;
					_isUpdateDownloadHandlerDeclared = true;
				}
						
				_appUpdate.DownloadUpdateAsync();
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(Language.strUpdateDownloadFailed, ex);
			}
		}
        #endregion
				
        #region Events
		private void DownloadUpdateProgressChanged(object sender, System.Net.DownloadProgressChangedEventArgs e)
		{
			prgbDownload.Value = e.ProgressPercentage;
		}
				
		private void DownloadUpdateCompleted(object sender, System.ComponentModel.AsyncCompletedEventArgs e)
		{
			try
			{
				btnDownload.Enabled = true;
				prgbDownload.Visible = false;
						
				if (e.Cancelled)
				{
					return ;
				}
				if (e.Error != null)
				{
					throw (e.Error);
				}
						
				if (MessageBox.Show(Language.strUpdateDownloadComplete, Language.strMenuCheckForUpdates, MessageBoxButtons.OKCancel, MessageBoxIcon.Warning) == System.Windows.Forms.DialogResult.OK)
				{
					Shutdown.Quit(_appUpdate.CurrentUpdateInfo.UpdateFilePath);
					return ;
				}
				else
				{
					File.Delete(_appUpdate.CurrentUpdateInfo.UpdateFilePath);
				}
			}
			catch (Exception ex)
			{
				Runtime.MessageCollector.AddExceptionMessage(Language.strUpdateDownloadCompleteFailed, ex);
			}
		}
        #endregion
	}
}
